nvbench_helper/nvbench_helper/device_side_benchmark.cuh

```diff
@@ -0,0 +1,43 @@
+// SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <cuda/cmath>
+#include <cuda/ptx>
+#include <cuda/std/cstdint>
+#include <cuda/std/cstring>
+#include <cuda/utility>
+
+template <typename T>
+__device__ __forceinline__ static T generate_random_data()
+{
+  constexpr auto size = cuda::ceil_div(sizeof(T), sizeof(uint32_t));
+  uint32_t data[size];
+  for (int i = 0; i < size; i++)
+  {
+    data[i] = cuda::ptx::get_sreg_clock();
+  }
+  T ret;
+  ::cuda::std::memcpy(&ret, data, sizeof(T));
+  return ret;
+}
+
+__device__ static int device_var[16];
+
+template <typename T>
+__device__ __forceinline__ static void sink(T value)
+{
+  if (cuda::ptx::get_sreg_smid() == static_cast<uint32_t>(-1))
+  {
+    *reinterpret_cast<T*>(device_var) = value;
+  }
```

Comment:
Could we just do a fake read here, with something like:
```c++
  asm("" : "+r"(value));
  ```
  ? I have seen this in benchmarks from @ahendriksen

Reply:
> `asm("" : "+r"(value));`

That will ensure NVVM believes that the value is used, but from PTX -> SASS, this is still likely to be optimized out.

Reply:
Understood, thx!